Secure Channel Protocol

SanjoDeundiak opened the pull request:
1. Add dedicated struct for SecureChannel message type
2. Implement Close message
3. Implement credentials refresh
4. Use more compact CBOR representation of some Identity-related structs (arrays instead of maps)
5. Use full SHA256 as `Identifier` and `ChangeHash`
6. Add `data_type` field to `VersionedData` struct

SanjoDeundiak committed:
feat(rust): improve `Identity` `CBOR` structures

SanjoDeundiak committed:
feat(rust): update default noise protocol name string

SanjoDeundiak committed:
test(rust): update test vectors

SanjoDeundiak committed:
feat(rust): add secure channel protocol

SanjoDeundiak committed:
feat(rust): handle close and refresh credentials secure channel messages

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_state_machine.rs:
```diff
@@ -116,77 +116,144 @@ impl CommonStateMachine {
-                return Err(IdentityError::InvalidKeyType.into())
+            (None, None) => {}
+            _ => {
+                return Err(IdentityError::InvalidKeyData.into());
```
It would help debugging to have a logging line saying in which case we ended-up: `(Some(_), None)` or `(None, Some(_))`

SanjoDeundiak replied:
I changed the signature to avoid that case at all

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_state_machine.rs:
```diff
@@ -116,77 +116,144 @@ impl CommonStateMachine {
+                return Err(IdentityError::SecureChannelTrustCheckFailed.into());
+            }
+            debug!(
+                "Initiator checked trust policy for SecureChannel from: {}",
```
Can we pass the specific role here?

SanjoDeundiak replied:
That's a little bit problematic due to delegate macro you used. I'll remove the role from the message instead.

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/options.rs:
```diff
@@ -11,6 +11,12 @@ use core::fmt;
+/// This is the interval before a credential expiration when we'll query for a new credential.
+pub const DEFAULT_REFRESH_CREDENTIAL_TIME_GAP: Duration = Duration::from_secs(60);
+
+/// Minimal interval before
```
before? Should it say "Minimum duration between 2 credential refreshes" (to avoid too frequent refreshes I guess)

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/options.rs:
```diff
@@ -21,6 +27,8 @@ pub struct SecureChannelOptions {
     pub(crate) timeout: Duration,
+    pub(crate) min_refresh_credential_interval: Duration,
+    pub(crate) credential_refresh_time_gap: Duration,
 }
```
For consistency I propose to name that field `refresh_credential_time_gap`

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channels/secure_channels.rs:
```diff
@@ -113,7 +147,9 @@ impl SecureChannels {
             access_control.decryptor_outgoing_access_control,
-            options.credentials,
+            credentials,
+            options.min_refresh_credential_interval,
```
I wonder if we shouldn't pass a `SecureChannelTrustOptions` struct containing all the options for credentials and trust (and have `SecureChannelListenerOptions` and `SecureChannelOptions` contain that struct)

SanjoDeundiak replied:
That's an interesting idea. I'll explore it when refactoring TrustContext

etorreborre commented on implementations/rust/ockam/ockam_identity/tests/credentials_refresh.rs:
```diff
@@ -0,0 +1,373 @@
+};
+use ockam_node::Context;
+
+struct LocalCredentialsRetriever {
```
I wouldn't mind reading this code after the tests in order to have more context :-)
Also one option would be to have:

- a `LocalCredentialsRetriever` returning credentials with no issue, exposing its call counter and used in test 2 and test 3
- a `FailingLocalCredentialsRetriever` wrapping the `LocalCredentialsRetriever` just to add the failing behaviour based one counters

SanjoDeundiak replied:
My C++ background doesn't allow me to declare something after it's used :) I'll move it
But I don't see a point of having 2 implementations, looks like excess almost copy&paste

etorreborre replied:
There are already 2 implementations in the test code :-) ([L.49](https://github.com/build-trust/ockam/pull/6843/files#diff-3a42c5c069a583c53235d3eb1ddc5ed61966d334ce64ad9bcfda2667354e374eR49) and [L.105](https://github.com/build-trust/ockam/pull/6843/files#diff-3a42c5c069a583c53235d3eb1ddc5ed61966d334ce64ad9bcfda2667354e374eR105)).
My proposal is to wrap one implementation the other so that there's no duplication of behaviour:
```rust
struct FailingLocalCredentialsRetriever {
    credential_retriever: LocalCredentialsRetriever,
    fail_iteration: Option<u8>,
    failed_call_counter: Option<Arc<AtomicU8>>,
}
```

SanjoDeundiak replied:
@etorreborre the duplication was a mistake and is now removed :D

etorreborre commented on implementations/rust/ockam/ockam_identity/tests/credentials_refresh.rs:
```diff
@@ -0,0 +1,373 @@
+
+    ctx.sleep(Duration::from_secs(10)).await;
+
+    assert_eq!(call_counter1.load(Ordering::Relaxed), 4);
```
I would be good to have a comment to tell why those values should be `4` and `3`.

etorreborre commented on implementations/rust/ockam/ockam_identity/tests/credentials_refresh.rs:
```diff
@@ -0,0 +1,373 @@
+        .unwrap()
+        .added();
+
+    assert!(added4 > added3);
```
Can we reorder the assertions as
```
assert!(added1 < added2);
assert_eq!(added2, added3);
assert!(added3 < added4);
```

etorreborre commented on implementations/rust/ockam/ockam_identity/tests/credentials_refresh.rs:
```diff
@@ -0,0 +1,373 @@
+    let identities_creation = identities.identities_creation();
+    let credentials = identities.credentials();
+
+    let authority = identities_creation
```
Since it takes a lot of vertical space we could have a function `create_identity(identities_creation)` returning an `Identifier`.

SanjoDeundiak replied:
The plan is to change the return type to `Identifier` in all existing public functions. I just didn't get to this change yet

etorreborre commented on implementations/rust/ockam/ockam_identity/tests/credentials_refresh.rs:
```diff
@@ -0,0 +1,373 @@
+        }
+    }
+
+    let call_counter2 = Arc::new(AtomicU8::new(0));
```
Since we're creating `client1` and `client2` in order it would be good to do the same for their trust contexts.
And possibly have a function creating a trust context just based on a credentials retriever.

SanjoDeundiak replied:
Renamed to client and server

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/handshake/handshake_worker.rs:
```diff
@@ -158,36 +163,46 @@ impl HandshakeWorker {
+
+        let min_credential_expiration = credentials
+            .iter()
+            .filter_map(|credential| credential.credential.get_versioned_data().ok())
```
It would be nice to have a function `expires_at()` which does all of this on `CredentialAndPurposeKey` directly

SanjoDeundiak replied:
I actually don't like either of the approaches. We should parse the data only once and then carry that information with some struct

etorreborre replied:
> We should parse the data only once and then carry that information with some struct

Yes I'm all for it 🎸

SanjoDeundiak replied:
But I'm not doing it as a part of that PR :D

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -110,14 +206,76 @@ impl EncryptorWorker {
+        Ok(())
+    }
+
+    async fn schedule_credentials_refresh(&mut self, ctx: &Context, is_retry: bool) -> Result<()> {
```
It would be nice to have a textual explanation of the logic in this function

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -110,14 +206,76 @@ impl EncryptorWorker {
+                self.min_refresh_credential_interval
+            } else {
+                Duration::from_secs(
+                    *(min_credential_expiration
```
It seems that we can compute this quantity once and use it in the `if` condition and in that branch. 
Otherwise I don't fully understand the logic. If `min_credential_expiration < now + self.refresh_credential_time_gap`, doesn't that mean that we are past the time for refreshing the credential and we should then refresh now and not in `min_refresh_credential_interval`?

I also don't get the meaning of `is_retry`. What are we retrying with this boolean?

SanjoDeundiak replied:
This quantity is used only here and only in the branch that guarantees we don't go into negatives

SanjoDeundiak replied:
Yes, I think we can put 0 in the other branch here. However, I would like to avoid situation when the credential retriever returns an error really fast, and we end up calling it very many times in a loop. In this case is_retry is true and we add the delay before calling retriever again

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -91,14 +120,81 @@ impl EncryptorWorker {
+                Ok(credential) => credential,
+                Err(err) => {
+                    info!(
+                        "Credentials refresh failed for {} and is rescheduled in {}",
```
`in {} seconds` would be more precise

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/options.rs:
```diff
@@ -11,6 +11,12 @@ use core::fmt;
 use core::fmt::Formatter;
 use core::time::Duration;
 
+/// This is the interval before a credential expiration when we'll query for a new credential.
```
I know it might be obvious to you but it could be good to add a comment about why we need this kind of parameter.
My guess is that we want to make sure that we don't end-up with an expired credential.

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -110,14 +206,76 @@ impl EncryptorWorker {
+        if let Some(min_credential_expiration) = self.min_credential_expiration {
+            self.credential_refresh_event = None;
+
+            let mut credential_refresh_event =
```
We could create this event closer to L.274

etorreborre reviewed (dismissed):
I left some comments on ideas for refactoring the code and the tests.
More importantly I would appreciate more comments / explanation around the scheduling logic to understand all the various cases.

One possibility would be to extract that logic as a function of `now`, `min_refresh_credential_interval`, `refresh_credential_time_gap`, `is_retry`, returning a `Duration` to schedule the `DelayedEvent`. 
Then have some unit tests showing the various outcomes depending on combination of inputs.

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -91,33 +141,154 @@ impl EncryptorWorker {
+            )
+        };
+
+        let duration = if duration < self.min_credential_refresh_interval && is_retry {
```
That's a nitpick but I think that would read even better as 
```rust
// Avoid too many request to the credential_retriever, the refresh can't be sooner than
// self.min_credential_refresh_interval if it's a retry
let duration = 
    if is_retry {
        min(self.min_credential_refresh_interval, duration)
    } else { 
        duration 
    };

SanjoDeundiak replied:
You tricked me, should be `max` instead of `min` 😃

etorreborre replied:
yes, `max` :-)

etorreborre commented on implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs:
```diff
@@ -91,33 +141,154 @@ impl EncryptorWorker {
+
+        let now = now()?;
+
+        let duration = if min_credential_expiration < now + self.refresh_credential_time_gap {
```
Why not simply
```rust
let duration = max(min_credential_expiration - self.refresh_credential_time_gap - now, Duration::from_secs(0));
```

SanjoDeundiak replied:
Isn't that gonna panic if  self.refresh_credential_time_gap + now > min_credential_expiration?

etorreborre replied:
In that case we take a duration of `0`. What would panic?

SanjoDeundiak replied:
@etorreborre https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6237699b4dafeefc6fe2206869cfdf25

etorreborre replied:
I see thanks. It's a bad inherited habit of working with signed integers.

SanjoDeundiak committed:
feat(rust): add `VersionedData::data_type`. remove hash truncation

polvorin committed:
feat(elixir): identity and secure channel protocol updates

Implement credential refresh, close and data messages as
defined in the secure channel messaging protocol.
Adapt to CBOR encoding changes around identities.

polvorin committed:
fix(elixir): re-add the noise test vector
